src/components/Launcher.jsx

```diff
@@ -0,0 +1,188 @@
+import Autosuggest from 'react-autosuggest';
+import { rawIntent } from '../lib/intents'
+
+const INTENT_VERB = 'OPEN'
+const INTENT_DOCTYPE = 'io.cozy.suggestions'
+
+class Launcher extends Component {
+  constructor (props) {
+    super(props)
+    
+    this.state = {
+      query: '',
+      suggestions: []
+    }
+    
+    this.sourcesNode = null
+    this.sources = []
+  }
+  
+  componentDidMount () {
+    // The launcher has one or more sources that provide suggestions. These sources are iframes into other apps, provied by thee intent system.
+    // Since we need to call the sources whenever the query changes, we are taking manual control over the intent process.
+    window.addEventListener('message', this.onMessageFromSource)
+    
+    rawIntent(INTENT_VERB, INTENT_DOCTYPE)
+    .then(intent => {
+      const { services } = intent.attributes
+      if (!services) return
+      
+      services.forEach(service => {
+        // we're injecting the iframes, but they will stay invisible
+        const iframe = document.createElement('iframe')
+        const url = service.href
+        iframe.setAttribute('src', url)
+        iframe.style.display = 'none'
+        this.sourcesNode.appendChild(iframe)
+        const serviceOrigin = url.split('/', 3).join('/')
+        
+        this.sources.push({
+          slug: service.slug,     // can be used to show where a suggestion comes from
+          origin: serviceOrigin,
+          id: intent._id,
+          ready: false,
+          window: null,           // will hold a reference to the window we're sending messages to
+          resolve: null           // a reference to a function to call when the source sends suggestions
+        })
+      })
+    })
+  }
+  
+  onMessageFromSource = (event) => {
+    // this re-implements a subset of injectService found in lib/intents, though only the part that are useful for suggestions
+    const source = this.sources.find(source => source.origin === event.origin)
+    
+    if (!source) return
+    
+    if (event.data.type === `intent-${source.id}:ready`) {
+      source.ready = true
+      source.window = event.source
+      
+      source.window.postMessage({}, event.origin)
+    }
+    else if (event.data.type === `intent-${source.id}:data` && source.resolve) {
+      source.resolve({
+        id: source.id,
+        suggestions: event.data.suggestions
+      })
+      source.resolve = null
+    }
+    else {
+      console.log('unhandled message:', event)
+    }
+  }
+  
+  onChange = (event, { newValue }) => {
+    this.setState({
+      query: newValue
+    })
+  }
+  
+  onSuggestionsFetchRequested = async ({ value }) => {
+    const promises = []
+    
+    this.sources.filter(source => source.ready).forEach(source => {
+      const promise = new Promise(resolve => {
+        source.resolve = resolve
+        source.window.postMessage({ query: value }, source.origin)
+      })
+      promises.push(promise)
+    })
```

Comment:
This is my favorite. Instead of creating an empty array then a `forEach` that `push` something, why don't we use a beautiful declarative `map` function:

```js
const promises = this.sources.filter(source => source.ready).map(source => 
  new Promise((resolve) => {
    source.resolve = resolve
    source.window.postMessage({ query: value }, source.origin)
  })
)
``` 

Is that more explicit to express the **what** instead of the **how**?

Reply:
Recommended read: https://tylermcginnis.com/imperative-vs-declarative-programming/